File: scala-lib/src/main/scala/org/wandledi/scala/ElementImpl.scala
package org.wandledi.scala

import org.xml.sax.Attributes
import org.wandledi.Attribute
import org.wandledi.Scroll
import org.wandledi.Selector
import org.wandledi.Spell
import org.wandledi.ElementForeachImpl
import org.wandledi.Plan
import org.wandledi.spells.InsertionIntent
import org.wandledi.spells.ReplacementIntent
import org.wandledi.spells.StringTransformation
import org.wandledi.spells.TransformedAttribute

class ElementImpl(
  aSelector: Selector,
  aScroll: Scroll)
extends org.wandledi.ElementImpl(aSelector, aScroll) with Element {

  def text_=(value: String) {
    text.setContent(value)
  }
  def text: TextContent = new TextContentImpl(this)

  def foreachIn[T](items: Iterable[T], reduceBefore: Boolean = false)
      (fun: (SelectableElement, T) => Unit)(implicit context: Selectable = null) {
    import collection.JavaConversions._
    val plan = new Plan[T] {
      def execute(e: org.wandledi.SelectableElement, item: T): Unit = {
        val se = new SelectableElementImpl(e.getSelector, e.getScroll, e.getScroll)
        def ex = fun(se, item)
        if (context != null) context.using(se)(ex)
        else ex
      }
    }
    val foreach = new ElementForeachImpl(this, items, reduceBefore)
    foreach.apply(plan)
  }

  def foreachWithIndexIn[T](items: Iterable[T], reduceBefore: Boolean = false)
      (fun: (SelectableElement, T, Int) => Unit)(implicit context: Selectable = null) {
    import collection.JavaConversions._
    val plan = new Plan[T] {
      def execute(e: org.wandledi.SelectableElement, item: T): Unit = {
        val se = new SelectableElementImpl(e.getSelector, e.getScroll, e.getScroll)
        def ex = fun(se, item, index)
        if (context != null) context.using(se)(ex)
        else ex
      }
    }
    val foreach = new ElementForeachImpl(this, items, reduceBefore)
    foreach.apply(plan)
  }

  def changeAttribute(name: String, change: (String) => String) {
    changeAttribute(name, new StringTransformation {
      def transform(value: String) = change(value)
    })
  }

  def changeAttributes(attr: (String, (String) => String)*) {
    val changes = attr.map { case (name, change) =>
      new TransformedAttribute(name, new StringTransformation {
          def transform(value: String) = change(value)
        })
    }
    changeAttributes(changes: _*)
  }

  def setAttributes(attr: (String, String)*) {
    setAttributes(attr.map(t => new Attribute(t._1, t._2)): _*)
  }

  def includeFile(file: String)(magic: => Unit)(implicit context: Selectable) {
    val scroll = new Scroll
    val sel = new SelectableImpl(scroll)
    context.using(sel)(magic)
    includeFile(file, scroll)
  }

  def insert(atEnd: Boolean)(insertion: (Spell) => Unit) {
    val intent = new InsertionIntent {
      def insert(parent: Spell) {
        insertion(parent)
      }
    }
    insert(atEnd, intent)
  }

  def insert(atEnd: Boolean = false, insertion: xml.NodeSeq) {
    insert(insertion.toString, atEnd)
  }

  def replace(contentsOnly: Boolean, replacement: (String, Attributes, Spell) => Unit) {
    val intent = new ReplacementIntent {
      def replace(label: String, attributes: Attributes, parent: Spell) {
        replacement(label, attributes, parent)
      }
    }
    replace(contentsOnly, intent)
  }

  def replace(contentsOnly: Boolean)(replacement: (String, Attributes) => xml.NodeSeq) {
    val intent = new ReplacementIntent {
      def replace(label: String, attributes: Attributes, parent: Spell) {
        parent.writeString(replacement(label, attributes).toString, true)
      }
    }
    replace(contentsOnly, intent)
  }

  def replace(contentsOnly: Boolean, replacement: xml.NodeSeq) {
    replace(contentsOnly, replacement.toString)
  }
}

File: core/src/main/java/org/wandledi/ElementImpl.java
package org.wandledi;

import java.util.Collection;
import org.wandledi.spells.*;
import org.wandledi.wandlet.Response;
import org.xml.sax.Attributes;


/**
 *
 * @author Markus Kahl
 */
public class ElementImpl implements Element {

    protected Scroll scroll;
    protected Selector selector;

    public ElementImpl(Selector selector, Scroll scroll) {
        this.scroll = scroll;
        this.selector = selector;
    }

    public Selector getSelector() {
        return selector;
    }

    public Scroll getScroll() {
        return scroll;
    }

    public TextContent getText() {
        return new TextContentImpl(this);
    }

    public ChargedElement max(int charges) {
        return new ChargedElement(selector, scroll, charges);
    }

    public LateElement at(int offset) {
        return new LateElement(selector, scroll, offset);
    }

    public void cast(Spell spell) {
        scroll.addSpell(selector, spell);
    }

    public void setAttribute(String name, String value) {
        cast(new AttributeTransformation(new Attribute(name, value)));
    }

    public void setAttributes(Attribute... attributes) {
        cast(new AttributeTransformation(attributes));
    }

    public void changeAttribute(final String name, final String value) {
        cast(new AttributeTransformation(new TransformedAttribute(name, new StringTransformation() {
            public String transform(String input) {
                return value.replace("$val", input);
            }

            public String toString() {
                return String.format(
                    "StringTransformation(change: '%s' -> '%s')",
                    name, value);
            }
        })));
    }

    public void changeAttribute(String name, StringTransformation transformation) {
        cast(new AttributeTransformation(new TransformedAttribute(name, transformation)));
    }

    public void changeAttributes(TransformedAttribute... attributes) {
        cast(new AttributeTransformation(attributes));
    }

    public void removeAttribute(String name) {
        cast(new AttributeTransformation(name));
    }

    public void clone(int times) {
        cast(new Duplication(times));
    }

    public void includeFile(String name) {
        cast(new Inclusion(name));
    }

    public void includeFile(String name, Scroll scroll) {
        cast(new Inclusion(name, scroll));
    }

    public void includeFile(Response response) {
        cast(new Inclusion(response.getFile(), response.getScroll()));
    }

    public void insert(boolean atEnd, InsertionIntent intent) {
        cast(new Insertion(atEnd, intent));
    }

    public void insert(String content) {
        insert(content, false);
    }

    public void insertLast(String content) {
        insert(content, true);
    }

    public void insert(final String content, final boolean atEnd) {
        InsertionIntent intent = new InsertionIntent() {
            public void insert(Spell parent) {
                parent.writeString(content, true);
            }

            public String toString() {
                return String.format("InsertionIntent(content: %s)", content);
            }
        };
        cast(new Insertion(atEnd, intent));
    }

    public void replace(boolean contentsOnly, ReplacementIntent intent) {
        cast(new Replacement(intent, contentsOnly));
    }

    public void replace(final boolean contentsOnly, final String content) {
        ReplacementIntent intent = new ReplacementIntent() {
            public void replace(String label, Attributes attributes, Spell parent) {
                parent.writeString(content, true);
            }

            public String toString() {
                return String.format("ReplacementIntent(content: %s)", content);
            }
        };
        cast(new Replacement(intent, contentsOnly));
    }

    public void truncate(int depth) {
        cast(new Truncate(depth));
    }

    public void reduce() {
        cast(new Reduction());
    }

    public void extract(Selector target) {
        cast(new Extraction(target));
    }

    public <T> ElementForeach<T> foreachIn(Collection<T> collection) {
        return new ElementForeachImpl<T>(this, collection);
    }

    public <T> ElementForeach<T> foreachIn(Collection<T> collection, boolean reduceBefore) {
        return new ElementForeachImpl<T>(this, collection, reduceBefore);
    }

    public void hide() {
        cast(new Invisibility());
    }
}
